source/server/admin/pprof_handler.cc

```diff
@@ -0,0 +1,204 @@
+                                            Buffer::Instance& response, AdminStream&) {
+  if (!kHasTcMalloc) {
+    return Envoy::Http::Code::NotImplemented;
+  }
+
+  perftools::profiles::Profile profile;
+
+  absl::flat_hash_set<void*> location_by_pointer;
+  absl::flat_hash_map<std::string, uint64_t> index_by_string;
+  absl::flat_hash_map<std::string, uint64_t> id_by_function;
+
+  auto get_string_index = [&index_by_string, &profile](absl::string_view str) {
+    auto it = index_by_string.find(str);
+    if (it == index_by_string.end()) {
+      it = index_by_string.emplace(str, profile.string_table_size()).first;
+      profile.add_string_table(std::string(str));
+    }
+    return it->second;
+  };
+
+  auto get_function_id = [&id_by_function, &profile, &get_string_index](absl::string_view name) {
+    auto it = id_by_function.find(name);
+    if (it == id_by_function.end()) {
+      auto id = id_by_function.size() + 1;
+      it = id_by_function.emplace(name, id).first;
+      auto* function = profile.add_function();
+      function->set_id(id);
+      function->set_name(get_string_index(name));
+    }
+    return it->second;
+  };
+
+  ABSL_ATTRIBUTE_UNUSED auto get_location_id = [&location_by_pointer, &profile,
+                                                get_function_id](void* pointer) {
+    uint64_t address = reinterpret_cast<uint64_t>(pointer);
+    if (!location_by_pointer.contains(pointer)) {
+      location_by_pointer.insert(pointer);
+      auto* location = profile.add_location();
+      location->set_id(address);
+      location->set_address(address);
+
+      char tmp_buffer[1024];
+      if (absl::Symbolize(pointer, tmp_buffer, sizeof(tmp_buffer))) {
+        auto* line = location->add_line();
+        line->set_function_id(get_function_id(tmp_buffer));
+      }
+    }
+    return address;
+  };
+
+  auto set_value_type = [&get_string_index](perftools::profiles::ValueType* value_type,
+                                            absl::string_view type, absl::string_view unit) {
+    value_type->set_type(get_string_index(type));
+    value_type->set_unit(get_string_index(unit));
+  };
+
+  // The first string must be empty.
+  get_string_index("");
+
+  set_value_type(profile.mutable_period_type(), "space", "bytes");
+  profile.set_default_sample_type(get_string_index(""));
+  set_value_type(profile.add_sample_type(), "inuse_objects", "count");
+  set_value_type(profile.add_sample_type(), "inuse_space", "bytes");
+
+#if defined(TCMALLOC)
+  auto snapshot = tcmalloc::MallocExtension::SnapshotCurrent(tcmalloc::ProfileType::kHeap);
+  profile.set_period(snapshot.Period());
+  snapshot.Iterate([&](const tcmalloc::Profile::Sample& sample) {
+    auto* pb_sample = profile.add_sample();
+    pb_sample->add_value(sample.count);
+    pb_sample->add_value(sample.sum);
+    for (int i = 0; i < sample.depth; ++i) {
+      pb_sample->add_location_id(get_location_id(sample.stack[i]));
```

Comment:
I wonder if some c++ libraries to generate profile protos from tcmalloc snapshots exist somewhere out there.  Having to maintain our own version within envoy doesn't seem great.

My other comment is the usual request for tests, but it would be good to see if it would be possible to minimize the amount of code involved and its complexity before we go too far down the path of adding tests.

Reply:
I haven't found any unfortunately.

Reply:
Yeah, it seems like there isn't one.  I think it's worth having this version until at least something more centrally maintained exists.  What experience do you have with the output produced by this code?  Have you been using something like it internally for some time?

I think that as for tests, the minimal version would be to ensure that accessing this admin handler produces some output and doesn't crash the proxy.  I assume there's some test file that exercises other existing admin handlers to which you can add the minimal test suggested above.